Refactor SSL engine and add optional GnuTLS support

paullouisageneau opened the pull request:
This PR refactors SSL handling to hide OpenSSL specifics and introduces an option to compile against GnuTLS instead of OpenSSL.

The motivation is to ease integration by allowing to avoid different TLS backends in a single project as well as potential license issues with OpenSSL (the OpenSSL license is an Apache License 1.0 which is incompatible with the GPL).

GnuTLS support can be enabled with the compilation flag `TORRENT_USE_GNUTLS`. SSL code is made generic where possible (for both trackers and SSL torrents) and SSL tests are modified accordingly.

The GnuTLS boost.asio wrapper is imported from [paullouisageneau/boost-asio-gnutls](https://github.com/paullouisageneau/boost-asio-gnutls) as submodule.

This is based on the preliminary work of @alexandre-janniaux and implements https://github.com/arvidn/libtorrent/issues/2858

alexandre-janniaux commented:
Hi, thank you for integrating my work! As they were WIP, I can rewrite my commits or you can rewrite and author them to your name with an additional `Co-authored-by` if you prefer rewriting them yourself.

arvidn commented:
This is quite a patch. I think there is quite a lot of interest in being able to use GnuTLS instead of OpenSSL. I'm a little bit hesitant to put the GnuTLS stream wrapper in libtorrent itself. It would seem more reasonable to either contribute it to boost.asio, or put it in a separate repository that libtorrent could pull in.

That said, obviously there would still need to be new abstractions in libtorrent itself.

Do you know if boost.asio has unit tests for the openssl stream? If so, hooking this stream wrapper into that test would probably make a lot of sense too.

paullouisageneau commented:
@alexandre-janniaux I'll rewrite and author them as you suggest!

@arvidn I'm not sure about contributing it to boost.asio for now, even if it completely covers the use case here, it is not a full featured wrapper. However putting it in another repository and only letting the libtorrent abstractions here makes sense. It could still be merged in boost.asio one day anyway given enough interest and development.

I'll start working on moving the wrapper and see about boost.asio's unit tests. In the meantime feel free to comment on the rest.

alexandre-janniaux commented on include/libtorrent/ssl/context.hpp:
```diff
@@ -165,9 +165,8 @@ class TORRENT_EXTRA_EXPORT context : public boost::asio::ssl::context
-		using namespace std::placeholders;
 		m_impl->expected_dn = std::move(name);
-		set_verify_callback(std::bind(&context::verify_dn, this, _1, _2), ec);
+		set_verify_callback(std::bind(&context::verify_dn, this, std::placeholders::_1, std::placeholders::_2), ec);
```
This could probably use some local `use std::placeholders` !

paullouisageneau replied:
@alexandre-janniaux That's right, I used it at first but eventually reverted because it caused an ambiguity with `boost::placeholders` in some of the CI environment (for instance the macOS build on travis)...

paullouisageneau commented:
@arvidn I've moved the GnuTLS wrapper in the independant repository [paullouisageneau/boost-asio-gnutls](https://github.com/paullouisageneau/boost-asio-gnutls) as you suggested. I added unit tests equivalent to the ones of asio/ssl, which are compile-time tests, I'm looking at the proper way to add runtime tests also.

arvidn commented on include/libtorrent/ssl/context.hpp:
```diff
@@ -0,0 +1,279 @@
+using base::stream_base;
+#endif
+
+class TORRENT_EXTRA_EXPORT context : public base::context
```
It would seem more appropriate to just make `context` a typedef to either the boost asio one or the gnutls one. As far as I can tell, all OpenSSL/GnuTLS specific code could still be done in free functions (ideally implemented in a .cpp file)

paullouisageneau replied:
A significant amount of code can be moved to .cpp files for sure. 
Free functions would have a drawback I think: I see no way to implement generic functions for `set_servername_callback` (because you need to keep the generic callback somewhere and openssl takes a raw function pointer) and `get_context` because you need to retrieve the wrapped context somehow).

arvidn replied:
but boost.asio's `ssl::context` already keeps that function object. You just have to do the same thing in the gnutls counterpart, no?

paullouisageneau replied:
It does for `set_verify_callback`, but it has no support for servername callback sadly.

arvidn commented on include/libtorrent/ssl/context.hpp:
```diff
@@ -0,0 +1,279 @@
+#endif
+
+private:
+	static void global_init();
```
this `global_init()` function doesn't seem all that useful. Couldn't you just do the openSSL initialization in a global constructor, as it used to be? Now it looks like you run it every time a `context` is created which happens multiple times. It also doesn't seem to benefit from being a member function here.

paullouisageneau replied:
The init was done every time a torrent context was created before: https://github.com/arvidn/libtorrent/blob/48627d7dfe47fb50a99fbd429de25bebec129dfc/src/torrent.cpp#L1636
I'll move it in a global constructor just like the cleanup is in a global destructor.

arvidn replied:
oh, I see. I think I confused it with how I used to initialize openssl a long time ago, where it was in the same global static-storage duration object that does the cleanup ( https://github.com/arvidn/libtorrent/blob/master/src/session_impl.cpp#L172 ). It's possible I moved it to be done more often for a reason. It could be risky to call it over and over again though

arvidn commented on include/libtorrent/ssl/context.hpp:
```diff
@@ -0,0 +1,279 @@
+	}
+
+#if defined TORRENT_USE_OPENSSL
+	error_code set_servername_callback(std::function<bool(stream_base& s, std::string name)> cb, error_code& ec)
```
does this not have an implementation for GnuTLS?
Either way, this used to be a free function with the implementation hidden in a .cpp file, I think it's desirable to put everything that needs to include an openssl header in a .cpp file. I at least had attempted to do so previously, with the `openssl.cpp`

paullouisageneau replied:
This is already implemented in the GnuTLS wrapper itself (I put it there since for GnuTLS, SNI is not an extension API, it's in the core lib).
I'll put everything I can in .cpp files.

arvidn replied:
Whether it's in the core API or not, the API *is* a callback still, right? I don't think it's important that callback itself is generic. Or rather, it's not clear to me the additional abstraction is worth making this callback SSL-library-agnostic.

paullouisageneau replied:
I think you are right, I'll reduce the abstraction by having two versions of the callback.

arvidn commented on src/openssl.cpp:
```diff
@@ -1,82 +0,0 @@
-
-#if OPENSSL_VERSION_NUMBER >= 0x90812f
-
-void openssl_set_tlsext_servername_callback(SSL_CTX* ctx
```
I haven't thought this through all the way, but it seems like you could keep these functions and just make them take an `lt::ssl::context&` instead of the openssl specific type.

arvidn replied:
oh, I guess the callback function signature is still openssl specific. If there isn't a GnuTLS implementation for this yet though, maybe that's OK for a first pass.

paullouisageneau replied:
Yes, that's exactly the problem, `SSL*` is OpenSSL specific, GnuTLS uses opaque handles and the native callback signatures are totally different.

arvidn commented on include/libtorrent/ssl/context.hpp:
```diff
@@ -0,0 +1,279 @@
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+		~cleanup() { CRYPTO_cleanup_all_ex_data(); }
```
I would prefer this code also lives in a .cpp file. `session_impl.cpp` is probably *not* the best place, but some counterpart to the previous `openssl.cpp` would make sense I think

arvidn commented on include/libtorrent/ssl/stream.hpp:
```diff
@@ -0,0 +1,217 @@
+};
+
+template<typename NextLayer>
+class stream : public base::stream<NextLayer>, public stream_base
```
is there a reason `stream` cannot be a typedef to either `boost::asio::ssl::context` or the gnutls counterpart?

paullouisageneau replied:
The only reason is the implementation of `get_context` in that case. However, I think it's doable by storing the reference to it in ex data of the corresponding `SSL_CTX`.

arvidn commented on src/openssl.cpp:
```diff
@@ -1,82 +0,0 @@
-// all of OpenSSL causes warnings, so we just have to disable them
-#include "libtorrent/aux_/disable_warnings_push.hpp"
-
-void openssl_set_tlsext_hostname(SSL* s, char const* name)
```
I think having this free function here (or in some .cpp file) is really nice. It hides the openssl/gnutls includes, to not leak into a dependency passed on to clients of libtorrent. It keeps the places where the warnings are disabled to a single place, because of these openssl macros.

There could be another version of this that take whatever the native handle is for a GnuTLS connection. the counterpart to `SSL*`.

paullouisageneau replied:
The set hostname function per se is no problem. If I manage to implement the callbacks with free functions then this one will follow.

arvidn commented on src/socket_type.cpp:
```diff
@@ -114,33 +105,17 @@ namespace aux {
-		if (ctx)
-		{
-			aux::openssl_set_tlsext_servername_callback(ctx, nullptr);
-			aux::openssl_set_tlsext_servername_arg(ctx, nullptr);
```
I have a feeling clearing this callback may be important

arvidn commented on src/socket_type.cpp:
```diff
@@ -88,23 +85,17 @@ namespace aux {
-			boost::asio::ssl::rfc2818_verification(hostname), ec); \
-		ssl = s.get<ssl_stream<t>>()->native_handle(); \
-		ctx = SSL_get_SSL_CTX(ssl); \
+		s.get<ssl_stream<t>>()->set_host_name(hostname, ec); \
```
I think it would be nice if the gnutls classes were compatible with boost.asio's ones as much as possible. iirc, asio actually has this verify callback with a generic signature, not dependent on OpenSSL

paullouisageneau replied:
The gnutls classes are compatible, for instance the signature is the same for `set_verify_callback`. The issue is that the `verify_context` is just a dummy container for an OpenSSL specific type.
However, I see now that I could add a `gnutls::rfc2818_verification` for this case.

arvidn commented on src/socket_type.cpp:
```diff
@@ -88,23 +85,17 @@ namespace aux {
-		SSL* ssl = nullptr;
-		SSL_CTX* ctx = nullptr;
-
 		switch(s.type())
```
one relatively big change in master is that `socket_type` now behaves more like a `boost::variant`, so instead of these switch statements and saving off SSL pointers, there are visitors. So I think this could be done a bit cleaner in the most recent `master` branch

arvidn replied:
specifically, where the concrete type of the native handles don't necessarily need to be mentioned, but just passed in to free functions that have two implementations, one for openssl and one for GnuTLS

paullouisageneau replied:
I see, that'd be nice. I'm not 100% sure this is doable with free functions for now because of the callbacks but I'll look into it.

arvidn commented on src/ssl/context.cpp:
```diff
@@ -0,0 +1,287 @@
+#endif
+}
+
+bool context::verify_dn(std::string const& expected, bool preverified, boost::asio::ssl::verify_context& ctx) {
```
This function is specific to running bittorrent over SSL. I don't think it belongs as a member function of the SSL context. I think it would make more sense for it to be a free function (or probably two free functions, on for OpenSSL and one for GnuTLS). But I would think `torrent.cpp` is a pretty reasonable place for it still.

paullouisageneau replied:
I'll move them there then.

arvidn commented on src/ssl/context.cpp:
```diff
@@ -0,0 +1,287 @@
+namespace {
+
+#if DEBUG_SSL_VERIFY
+void debug_log(char const* fmt, ...)
```
this used to print logsas alerts, and make them available to clients. I see the verification functions used to be member functions on `torrent`, was the alert the only thing it needed the torrent object for?

paullouisageneau replied:
Yes it was the only thing. So the alerts were only available in debug mode by design?

arvidn replied:
it was enabled unless `TORRENT_DISABLE_LOGGING` was defined, which actually makes it available by default (but only when enabled in the `alert_mask`). I don't really have strong opinions about preserving this logging though. I imagine I added it while trouble shooting certs issues in SSL swarms

arvidn commented on src/ssl/stream.cpp:
```diff
@@ -0,0 +1,54 @@
+void stream_base::init_base()
+{
+    if (ex_index < 0) ex_index = SSL_get_ex_new_index(0, NULL, NULL, NULL, NULL);
+    SSL_set_ex_data(native_handle(), ex_index, this); // expected by contex
```
what does this do? I think this is a new call

paullouisageneau replied:
This is OpenSSL's way of setting a user pointer. It is used in `ssl::context` to retrieve a pointer to the `stream_base` from the `SSL*` native handle in the servername callback.

arvidn commented on src/torrent.cpp:
```diff
@@ -1530,141 +1527,27 @@ bool is_downloading_state(int const st)
 		// create the SSL context for this torrent. We need to
 		// inject the root certificate, and no other, to
 		// verify other peers against
 #if BOOST_VERSION >= 106400
```
in the latest `master`, boost older than 1.64 is no longer supported, so this was simplified a bit

arvidn commented on src/torrent.cpp:
```diff
@@ -1675,54 +1558,28 @@ bool is_downloading_state(int const st)
 		// the verification function verifies the distinguished name
 		// of a peer certificate to make sure it matches the info-hash
 		// of the torrent, or that it's a "star-cert"
-		ctx->set_verify_callback(std::bind(&torrent::verify_peer_cert, this, _1, _2), ec);
```
I think it would be nicer if the gnutls context would have `set_verify_callback()` and `verify_dn` would be a free function (or two free functions, but only one would be in use at a time).

I think it's an improvement to decouple it from being a `torrent` member function as you have done, but I think it should be decoupled from the context too.

paullouisageneau replied:
OK I'll see what I can do.

arvidn commented:
one part of this patch that I think could land in `master` right away is the `TORRENT_USE_OPENSSL` and `TORRENT_USE_SSL` macro change. However, what do you think of making `TORRENT_USE_SSL` be defined to 0 or 1 and tested like `#if TORRENT_USE_SSL` instead of `#ifdef`?

I think in general that's a better practice, at least for internal configuration macros, as the compiler can warn on misspelled macros.

paullouisageneau commented:
@arvidn I've refactored everything into free functions, I think a managed to do something clean and generic enough. I also changed `TORRENT_USE_SSL` to be defined to be either 0 or 1 as suggested. Would you like me to extract this part and open a new PR targeting master?

arvidn commented on include/libtorrent/proxy_base.hpp:
```diff
@@ -124,6 +124,18 @@ class proxy_base
 		m_sock.async_write_some(buffers, handler);
 	}
 
+#if BOOST_VERSION >= 106600 && !defined TORRENT_BUILD_SIMULATOR
```
I imagine I would have to add `async_wait` to uTP connections as well. and I really should add it to libsimulator (but I suppose the simulator doesn't support SSL right now anyway).

I take it this is used by the gnutls wrapper?

paullouisageneau replied:
Yes it is. I observed strange behaviors with null buffers, so since the are now deprecated anyway, I used the recommended `async_wait` which worked fine.

arvidn commented on include/libtorrent/utp_stream.hpp:
```diff
@@ -493,6 +518,33 @@ struct TORRENT_EXTRA_EXPORT utp_stream
 	}
 
+#if BOOST_VERSION >= 106600
+	// Compatiblity with the async_wait method introduced in boost 1.66
```
oh, I see, you added it :)

arvidn replied:
I think this function, along with the `proxy_base` support for `async_wait` is a good candidate for a separate PR (that can land ahead of this big one, to make it just a little bit smaller)

arvidn commented on include/libtorrent/utp_stream.hpp:
```diff
@@ -420,11 +421,35 @@ struct TORRENT_EXTRA_EXPORT utp_stream
+		TORRENT_ASSERT(ret <= buf_size);
+		if(ret == 0)
+		{
+			ec = boost::asio::error::would_block;
```
this isn't quite right. if `write_some()` returns 0, it just means it couldn't send any packets *right now*. However, since you called `add_write_buffer()` already, the socket is remembering those pointers and will read from them later, when it *can* send more packets on the socket.

The normal behavior of a blocking write is that the socket/kernel will copy some of the data into its own send buffer, which later will be sent. If that send buffer is full, the call fails with `EWOULDBLOCK`. There isn't a send buffer in the uTP sockets though. Since they only support asynchronous writes, all they support is to remember where the user's buffers are, and read directly from them.

Have you seen this work? does the ssl unit test pass with the GnuTLS backend?

I can't think of a simple way of supporting this. I think we'd either need a send buffer, or have this never buffer anything but just consume any data that can be sent as packets *right now* (which most likely will be 0 most of the time)

paullouisageneau replied:
You are right, the `write_some` I implemented only supports non-blocking immediate writes due to the limitations you mentioned (similar to socket behavior with the non-blocking option set to true), which is OK as long as you don't mix it with asynchronous writes of course. I'll comment about that.
Luckily, it's only what I needed for the GnuTLS wrapper. It allows to make it much simpler as relying on async writes would require buffering data on the wrapper level.

arvidn commented on src/utp_stream.cpp:
```diff
@@ -1084,25 +1084,43 @@ std::size_t utp_stream::read_some(bool const clear_buffers)
 	return ret;
 }
 
+std::size_t utp_stream::write_some(bool const clear_buffers)
```
I think this patch (along with support for syncrhonous write) is also a good candidate for a separate PR that could land ahead of this one

arvidn commented:
I think this is looking good. I haven't reviewed everything yet, but what I like what I've seen so far.

In the interest of making it more manageable to review the core of this patch, I think it would be beneficial to break out some chunks.

The `TORRENT_USE_SSL` define is a good candidate I think. The broken-out patch should probably be before GnuTLS is introduced, so mostly a new internal name (`TORRENT_USE_SSL`) with the same external name (the one build systems define, `TORRENT_USE_OPENSSL`).

I point out two other candidates too, which seem somewhat independent

arvidn commented:
I just noticed this is targeting `RC_1_2`. I'm not sure that's possible, ABI wise. It might, but it would also create quite a headache merging this into master.

Do you think it's important it makes it into `RC_1_2`?
I try to limit changes to bug fixes, and simple changes.

paullouisageneau commented:
It's not important that it makes it into `RC_1_2`, therefore I'll target master and break it down into 3 parts as you suggest: `TORRENT_USE_SSL` renaming, `async_wait` and `read_some` methods, and eventually the actual GnuTLS support.

arvidn commented:
in `master`, there's a new abstraction to represent the `info-hash` of a torrent, with bittorrent v2 support, it may be both sha-1 and sha-256. In most cases it can be handled the same though.
https://ci.appveyor.com/project/arvidn/libtorrent/builds/32746464/job/y4i5agrgi5x2m1px#L225

paullouisageneau commented:
@arvidn Thanks for merging the prerequisites, I rebased on master and fixed the info_hash issue.

arvidn commented:
great! I'm hoping to do a more careful review later today. One thing I think might be a good idea is to add a build in `.travis.yml` with `crypto=gnutls`, to demonstrate it working. I imagine getting it to work on windows and MacOS might be a bit challenging, but a linux build would be doable I hope.

does that make sense?

paullouisageneau commented:
Yes, a Linux build should be doable, I'm adding it.

arvidn commented on CMakeLists.txt:
```diff
@@ -651,35 +652,67 @@ if(Iconv_FOUND)
-			TORRENT_SSL_PEERS
-			OPENSSL_NO_SSL2)
-	target_sources(torrent-rasterbar PRIVATE src/pe_crypto)
+if(gnutls OR NOT OPENSSL_FOUND)
```
I'm not quite following this. `gnutls` is set by the first call to `find_public_dependency(GnuTLS)`, is it?

If it's set and `OPENSSL_FOUND` is *not* set, then you call `find_public_dependency(GnuTLS)` again? Why does it need to be called twice?

arvidn replied:
are the first calls to `find_public_dependency(GnuTLS)` and `set_package_properties(GnuTLS, ...` lef-overs that aren't needed anymore?

paullouisageneau replied:
No, `gnutls` is actually the cmake option set by the user (disabled by default). So the logic is to use GnuTLS if either set by the user or if OpenSSL is not found.

paullouisageneau replied:
However the second call is a rebase mistake I think, it's indeed useless :facepalm:

arvidn commented on CMakeLists.txt:
```diff
@@ -651,35 +652,67 @@ if(Iconv_FOUND)
 if(OPENSSL_FOUND)
 	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_LIBCRYPTO)
-else(OPENSSL_FOUND)
+	target_sources(torrent-rasterbar PRIVATE src/pe_crypto)
```
`pe_crypto` implements the bittorrent "encryption"/obfuscation protocol. It uses diffie hellman and RC4. It doesn't require libcrypto, but I think it will use its routines if it's available. It's already being added below, if `encryption` is enabled. I don't think it needs to be added here, or in the else-clause

arvidn commented on Jamfile:
```diff
@@ -407,6 +411,43 @@ rule openssl-include-path ( properties * )
+	{
+		# on macOS, default to pick up gnutls from the homebrew installation
+		# brew install gnutls
+		GNUTLS_INCLUDE = /usr/local/opt/openssl/include ;
```
this include looks wrong. did you test this on MacOS?

paullouisageneau replied:
No, I don't have such a machine on hand. I'll try on travis when the Linux build works.

arvidn replied:
it probably shouldn't say "openssl" either way :)

arvidn replied:
I can try this on mac tomorrow (if I remember)

arvidn commented on include/libtorrent/ssl_stream.hpp:
```diff
@@ -34,14 +34,14 @@ POSSIBILITY OF SUCH DAMAGE.
 #include "libtorrent/error_code.hpp"
 #include "libtorrent/io_context.hpp"
-#include "libtorrent/aux_/openssl.hpp"
+#include "libtorrent/ssl.hpp"
```
should the `ssl.hpp` header really be included regardless of whether `TORRENT_USE_SSL` is 1?

arvidn commented on include/libtorrent/ssl_stream.hpp:
```diff
@@ -34,14 +34,14 @@ POSSIBILITY OF SUCH DAMAGE.
-#include "libtorrent/aux_/openssl.hpp"
+#include "libtorrent/ssl.hpp"
+
+#if TORRENT_USE_SSL
```
actually, it's not clear to me why you moved this down. if `TORRENT_USE_SSL == 0`, this whole file will turn into nothing, right?

paullouisageneau replied:
Indeed, I just moved it down because it was shorter than including `config.hpp` beforehand.

arvidn commented on src/hasher.cpp:
```diff
@@ -194,7 +193,7 @@ TORRENT_CRYPTO_NAMESPACE
 	hasher256& hasher256::operator=(hasher256 const& h) &
 	{
-		if (this == &h) return;
+		if (this == &h) return *this;
```
ouch

arvidn commented on src/session_impl.cpp:
```diff
@@ -452,51 +452,45 @@ void apply_deprecated_dht_settings(settings_pack& sett, bdecode_node const& s)
 
 		info_hash_t info_hash;
-		bool valid = aux::from_hex({servername, 40}, info_hash.v1.data());
+        bool valid = aux::from_hex({name.c_str(), 40}, info_hash.v1.data());
```
spaces instead of tabs here, for indentation

arvidn commented on src/ssl.cpp:
```diff
@@ -0,0 +1,257 @@
+    ca.size = unsigned(str.size());
+
+	// Warning: returns the number of certificates processed or a negative error code on error
+	int ret = gnutls_certificate_set_x509_trust_mem(c.native_handle(), &ca, GNUTLS_X509_FMT_PEM);
```
I thought it looked a bit weird that the `data` field needed to be null terminated, given that you pass in the length as well. So, I guess [this](https://www.gnutls.org/manual/html_node/Core-TLS-API.html#index-gnutls_005fcertificate_005fset_005fx509_005ftrust_005fmem) is the authoritative documentation.

I don't see it say anything about it needing to be 0-terminated. But I see `ca` is a *list* of certificates. How does it know how many certs are in the list? You pass in 1, but what stops the function from assuming there are more after? Are you sure it's not the list of certs that's null terminated? Say, the last `data` field is `nullptr` to indicate the end of the list.

If that is the case, you wouldn't have to copy the certificate above.

paullouisageneau replied:
Oh, that's true. Other gnutls x509 functions require it though, like [gnutls_certificate_set_x509_key_mem](https://www.gnutls.org/manual/html_node/Core-TLS-API.html#index-gnutls_005fcertificate_005fset_005fx509_005fkey_005fmem)

I'm pretty sure the fact that it can be a list in the case of PEM encoding only means you can list different CA in the same string.

arvidn replied:
I see. Right, I think I was in C++ mode thinking that there must be a reason it was a pointer instead of being passed by value or reference :)

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1383,87 @@ bool is_downloading_state(int const st)
 #endif
 
 #ifdef TORRENT_SSL_PEERS
+	void torrent::init_ssl(string_view cert)
```
it looks like `torrent::init_ssl()` moved up, so it no longer lines up with where the previous function was. This makes it a bit more difficult to see what changed in here. Would you mind moving it down to where it currently is in `master`?

arvidn replied:
i.e. move `torrent::init_ssl()` down right after `torrent::verify_peer_cert`

paullouisageneau replied:
Oh yes sorry!

paullouisageneau commented on src/ssl.cpp:
```diff
@@ -0,0 +1,257 @@
+#endif
+}
+
+void set_host_name(stream_handle_type s, std::string const& name, error_code& ec)
```
Yes you are right!

The underlying issue is that GnuTLS does not have the same distinction of context and stream, so the native types exposed by `gnutls::context` and `gnutls::stream` can't be used as their OpenSSL counterpart. Especially for switching contexts, since the `gnutls::context` is actually only a credentials handler, the rest is emulated by the wrapper.

paullouisageneau commented on include/libtorrent/ssl.hpp:
```diff
@@ -0,0 +1,182 @@
+using native_context_type = context::native_handle_type;
+using native_stream_type = stream_base::native_handle_type;
+using context_handle_type = context&;
+using stream_handle_type = stream_base&;
```
Very good point, I'm changing for pointers.

As you guessed in the next comment, the goal of the handle is not to make callbacks uniform, it is to be able to pass either the native handle or the wrapper itself the to the `ssl` free functions.

paullouisageneau commented on src/ssl.cpp:
```diff
@@ -0,0 +1,257 @@
+} // anonymous namespace
+#endif
+
+void set_server_name_callback(context_handle_type c, server_name_callback_type cb, void* arg, error_code& ec)
```
That would be great, sadly it's not possible since `SSL_CTX_set_tlsext_servername_callback` expects a C-style callback, not a C++ callable.

paullouisageneau commented on src/ssl.cpp:
```diff
@@ -0,0 +1,257 @@
+void set_server_name_callback(context_handle_type c, server_name_callback_type cb, void* arg, error_code& ec)
+{
+#if defined TORRENT_USE_OPENSSL
+#if OPENSSL_VERSION_NUMBER >= 0x90812f
```
OK great!

paullouisageneau commented on src/session_impl.cpp:
```diff
@@ -452,51 +452,45 @@ void apply_deprecated_dht_settings(settings_pack& sett, bdecode_node const& s)
 	// trying to connect to. The 40 first bytes in the name is expected to
 	// be the hex encoded info-hash
-	int servername_callback(SSL* s, int*, void* arg)
+	bool ssl_server_name_callback(ssl::stream_handle_type stream_handle, std::string const& name, void* arg)
```
OK, works for me!

paullouisageneau commented:
I fixed the Jamfile, the new travis build with `crypto=gnutls` seems to work.

arvidn commented on src/utp_stream.cpp:
```diff
@@ -651,10 +651,7 @@ void utp_socket_impl::issue_write()
-	// no more payload to send or if the congestion window
-	// is full and we can't send more packets right now
-	while (send_pkt());
+	write_some(false);
```
if you revert this, then `write_some()` wouldn't need to take the `bool` parameter, right?
it's not clear that reusing this function here is worth it. it's a single line of code, and a `bool` to disable two other lines of code in there.

paullouisageneau replied:
Sure, actually I only added it so it is symmetrical with `read_some`.

arvidn commented on test/test_ssl.cpp:
```diff
@@ -476,11 +469,11 @@ bool try_connect(lt::session& ses1, int port
 
 		std::printf("SNI: %s\n", name.c_str());
-		aux::openssl_set_tlsext_hostname(ssl_sock.native_handle(), name.c_str());
+		ssl::set_host_name(ssl::get_handle(ssl_sock), name, ec);
```
```suggestion
		ssl::set_host_name(ssl::get_handle(ssl_sock), name, ec);
		TEST_CHECK(!ec);
```

arvidn commented on test/test_ssl.cpp:
```diff
@@ -465,7 +458,7 @@ bool try_connect(lt::session& ses1, int port
 		std::string name = aux::to_hex(t->info_hash().v1);
 		std::printf("SNI: %s\n", name.c_str());
-		aux::openssl_set_tlsext_hostname(ssl_sock.native_handle(), name.c_str());
+		ssl::set_host_name(ssl::get_handle(ssl_sock), name, ec);
```
```suggestion
		ssl::set_host_name(ssl::get_handle(ssl_sock), name, ec);
		TEST_CHECK(!ec);
```

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,87 @@ bool is_downloading_state(int const st)
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#pragma clang diagnostic ignored "-Wused-but-marked-unused"
```
I think you also need:

```
#pragma clang diagnostic ignored "-Wold-style-cast"
#pragma clang diagnostic ignored "-Wzero-as-null-pointer-constant"
```

arvidn commented on src/ssl.cpp:
```diff
@@ -0,0 +1,219 @@
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+	SSL_set_tlsext_host_name(s, name.c_str());
```
I'm getting this warning:
```
../src/ssl.cpp:130:30: warning: cast from 'const char *' to 'char *' drops const qualifier [-Wcast-qual]
```

I think you also need:

```
#pragma clang diagnostic ignored "-Wcast-qual"
```

before this patch, all warnings are suppressed when using these openssl macros, by `#include "libtorrent/aux_/disable_warnings_push.hpp"`

paullouisageneau commented on src/session_impl.cpp:
```diff
@@ -61,7 +61,7 @@ POSSIBILITY OF SUCH DAMAGE.
 #include "libtorrent/aux_/disable_warnings_pop.hpp"
 
-#include "libtorrent/aux_/openssl.hpp"
+#include "libtorrent/ssl.hpp"
```
Oh, indeed. It seems I was quite inattentive when rebasing :sweat_smile:

paullouisageneau commented:
Hum, this seems platform-specific though, so it's probably linked to how error codes of the socket layer are handled. I'm having a look.

arvidn commented:
Here's what I know about the test failure:

`boost::asio::read()` fails ([here](https://github.com/arvidn/libtorrent/pull/4588/files#diff-ae80e01a9156b9c10eb22a8606c6fc57R520)) with error "Resource temporarily unavailable" (i.e. `EAGAIN`).

That's strange because the wrapper `boost::asio::read()` is supposed to keep reading until the requested number of bytes have been received, blocking. So I would expect it to *never* fail with this error.

arvidn commented:
looking into the gnutls `stream.hpp` wrapper. I notice that `gnutls_record_recv()` returns `GNUTLS_E_AGAIN` and you translate that to the error_code counterpart. What's strange about this is that the SSL socket (i.e. gnutls::stream) isn't supposed to have been set to non-blocking mode. (by default sockets are blocking in asio), so the call to `gnutls_record_recv` really should have blocked until some data become available. (but maybe all gnutls APIs are nob-blocking and it expects you to feed it more data in this case).

Looking around a bit in `stream.hpp`, I see a fair number of calls to `m_next_layer.non_blocking()`, which makes me wonder. Does your gnutls wrapper not support blocking mode?
it seems like it, and maybe it doesn't make much sense to support it since libtorrent itself doesn't use blocking mode. This test does though :/

arvidn commented:
it looks like `read_some()` might be missing some `gnutls_record_get_direction()` and `async_schedule()` in case you get `GNUTLS_E_AGAIN`. But I don't fully understand how the stream class works

arvidn commented on src/ssl.cpp:
```diff
@@ -0,0 +1,211 @@
+	SSL_set_SSL_CTX(s, c);
+	SSL_set_verify(s
+			, SSL_CTX_get_verify_mode(c)
+			, SSL_CTX_get_verify_callback(c));
```
sorry for nitpick; these two lines are indented one level too many

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
+#else
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tlsv12));
+#endif
+
```
In the existing `torrent::init_ssl()` there was some error checking here if `!ctx`, which you removed. It doesn't seem like it would make sense to check that here though, so it should be fine to remove it. If the `context` constructor fails, it will throw, to make `operator new` return `nullptr`.

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
+		// inject the root certificate, and no other, to
+		// verify other peers against
+#if BOOST_VERSION >= 106400
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tls));
```
this use `std::make_unique<>` now

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
+#if BOOST_VERSION >= 106400
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tls));
+#else
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tlsv12));
```
support for old boost is not necessary in `master`

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
+		// if all went well, set the torrent ssl context to this one
+		m_ssl_ctx = std::move(ctx);
+		// tell the client we need a cert for this torrent
+		alerts().emplace_alert<torrent_need_cert_alert>(get_handle());
```
these 4 lines used to be at the end of the function (after the `#if 0`-block)

arvidn commented on src/torrent.cpp:
```diff
@@ -1424,13 +1491,12 @@ bool is_downloading_state(int const st)
 				match = true;
 				// if we're logging, keep looping over all names,
 				// for completeness of the log
-				continue;
```
what happened to this `continue`?

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
+#ifndef TORRENT_DISABLE_LOGGING
+		std::string names;
+		bool match = false;
+#endif
```
these 4 lines used to be right before the loop with `sk_GENERAL_NAME_num`. It seems unnecessary to make the patch larger by moving them

paullouisageneau replied:
I put them here rather than putting them twice, since they are shared between OpenSSL and GnuTLS.

arvidn replied:
ah, of course. makes sense

arvidn commented on src/torrent.cpp:
```diff
@@ -1456,130 +1522,80 @@ bool is_downloading_state(int const st)
-#ifdef TORRENT_DISABLE_LOGGING
-				return true;
-#else
+#ifndef TORRENT_DISABLE_LOGGING
```
this was inverted, I don't think that's necessary, or necessarily an improvement

paullouisageneau commented:
> Looking around a bit in `stream.hpp`, I see a fair number of calls to `m_next_layer.non_blocking()`, which makes me wonder. Does your gnutls wrapper not support blocking mode?
> it seems like it, and maybe it doesn't make much sense to support it since libtorrent itself doesn't use blocking mode. This test does though :/

Thanks, I found the issue. The gnutls wrapper should work fine in blocking mode too, only the underlying socket is set to non-blocking mode for async operations. However, after the handshake I set the socket non-blocking by mistake. For some reason `test_ssl` still passed on my machine without hitting EAGAIN. It should be OK now.

arvidn commented:
> For some reason test_ssl still passed on my machine without hitting EAGAIN. It should be OK now.

Works on MacOS for me now too!

arvidn commented:
I think this looks good. A few things:

* There should probably be an entry in `ChangeLog` saying GnuTLS is supported
* The `Makefile` is used to build a tarball, I think it should pull in the files from the `deps/asiognutls` submodule as well, to make sure that makes it into a source distribution. Similar to how `deps/try_signal` is pulled in.

paullouisageneau commented:
@arvidn Great! I updated the Makefile to pull files from `deps/asio-gnutls`, added a line to the changelog, and squashed some commits. Is it expected that The Doozer MacOS build does not run?

arvidn commented:
yeah, the mac on doozer have had problems for a long time (I'm still hopeful it will come back to life, but eventually I will have to remove the doozer file)

arvidn commented on CMakeLists.txt:
```diff
@@ -651,35 +652,57 @@ if(Iconv_FOUND)
-	target_link_libraries(torrent-rasterbar PUBLIC OpenSSL::SSL)
-	target_compile_definitions(torrent-rasterbar
-		PUBLIC
-			TORRENT_USE_LIBCRYPTO
```
I just realized this was removed. It looks like it was redundant though, since it's added again below, if `OPENSSL_FOUND`

paullouisageneau replied:
I put it back here and remove it below.

arvidn commented on CMakeLists.txt:
```diff
@@ -651,35 +652,57 @@ if(Iconv_FOUND)
+			PURPOSE "Provides HTTPS support to libtorrent"
+	)
+	if(GNUTLS_FOUND)
+		target_link_libraries(torrent-rasterbar PUBLIC GnuTLS::GnuTLS)
```
it just occurred to me that `libgcrypt` is a dependency of `libgnutls`, isn't it? Unless I've misunderstood something, it might make sense to "enable" LibGcrypt when building in gnutls mode, like how `libcrypto` is "enabled" when building in openssl mode. This is a refinement that could be done later though, as it's not obvious whether it's a good idea or what the consequences might be.

paullouisageneau replied:
`libgcrypt` was indeed the cryptographic backend for `libgnutls`, however they changed for `nettle` in the last versions, so there is no guarantee the libs will be present together.

arvidn commented on src/torrent.cpp:
```diff
@@ -60,11 +60,8 @@ POSSIBILITY OF SUCH DAMAGE.
 #include <functional>
 
 #ifdef TORRENT_SSL_PEERS
+#include "libtorrent/ssl.hpp"
```
ssl.hpp is included twice in this file. Once conditionally (here), and then unconditionally a bit further down. I imagine this include is redundant.

arvidn commented on src/torrent.cpp:
```diff
@@ -1384,16 +1382,25 @@ bool is_downloading_state(int const st)
 #ifdef TORRENT_SSL_PEERS
-
-	bool torrent::verify_peer_cert(bool const preverified, boost::asio::ssl::verify_context& ctx)
+	bool torrent::verify_peer_cert(std::string const& expected, bool preverified, ssl::verify_context& ctx)
```
nit: `preverified` lost its `const`. it doesn't look like it needs to be mutable.

arvidn commented on src/torrent.cpp:
```diff
@@ -1479,37 +1538,20 @@ bool is_downloading_state(int const st)
-			| context::single_dh_use);
+		ctx->set_options(ssl::context::default_workarounds
+			| ssl::context::no_sslv2
+			| ssl::context::no_sslv3
```
SSLv3 is outdated an insecure by now I assume?

paullouisageneau replied:
Indeed, anyway this is somewhat redundant since now the method is set to `tls`.

arvidn commented on src/torrent.cpp:
```diff
@@ -1520,54 +1562,25 @@ bool is_downloading_state(int const st)
 		// of the torrent, or that it's a "star-cert"
-		ctx->set_verify_callback(std::bind(&torrent::verify_peer_cert, this, _1, _2), ec);
+		ctx->set_verify_callback(
+				std::bind(&torrent::verify_peer_cert, this, m_torrent_file->name(), _1, _2)
```
technically, since `verify_peer_cert` is still a member function now, `m_torrent_file->name()` doesn't need to be passed along here

paullouisageneau commented:
I addressed your few comments, thank you for the approval!

arvidn commented:
@paullouisageneau thanks! this is impressive work!

paullouisageneau commented:
@arvidn Thank you!
